[![Github Release](https://img.shields.io/badge/release-v2.1-brightgreen)](https://github.com/3UTR/DaPars2)
[![python Release](https://img.shields.io/badge/python-3.8-brightgreen)](https://www.python.org/downloads/)
[![numpy Release](https://img.shields.io/badge/numpy-1.22-brightgreen)](https://numpy.org/)
[![scipy Release](https://img.shields.io/badge/numpy-1.80-brightgreen)](https://scipy.org/)
[![R Release](https://img.shields.io/badge/R-3.6.3-brightgreen)](https://cran.r-project.org/)

# Dapars2 Modified

Modified from Dapars (v2.1)

## Usage

```
usage: DaPars2 [-h] -b gene_bed_file -s gene_symbol_annotation_file -w wigfiles [wigfiles ...] -f flagstat [flagstat ...] [-t threads] [-c first_100_3UTR_coverage_threshold] [--sample-list sample_list [sample_list ...]] [-o output_folder] [--treat-index TREAT_INDEX [TREAT_INDEX ...]]
               [--control-index CONTROL_INDEX [CONTROL_INDEX ...]] [--3UTR-Coverage-cutoff COVERAGE_CUTOFF] [--Num-least-in-treat NUM_LEAST_IN_TREAT] [--Num-least-in-control NUM_LEAST_IN_CONTROL] [--FDR-cutoff FDR_CUTOFF] [--PDUI-cutoff PDUI_CUTOFF]
               [--Fold-change-cutoff FOLD_CHANGE_CUTOFF]

DaPars2(Modified): Dynamics analysis of Alternative PolyAdenylation from multiple RNA-seq data.

optional arguments:
  -h, --help            show this help message and exit

base parameters:
  parameters used in DaPars2 base analysis

  -b gene_bed_file, --bed gene_bed_file
                        Input gene bed file, which is downloaded from UCSC Table browser (default: None)
  -s gene_symbol_annotation_file, --symbol gene_symbol_annotation_file
                        Input gene symbol annotation file, which is downloaded from UCSC Table browser (default: None)
  -w wigfiles [wigfiles ...], --wigfiles wigfiles [wigfiles ...]
                        wig files generated by: "bedtools genomecov -ibam *.bam -bga -split -trackline" (default: None)
  -f flagstat [flagstat ...], --flagstat-files flagstat [flagstat ...]
                        flagstat files generated by: "samtools flagstat *.bam", only pair-end support (default: None)
  -t threads, --threads threads
                        Number of threads (default: 4)
  -c first_100_3UTR_coverage_threshold, --first-100-3UTR-coverage first_100_3UTR_coverage_threshold
                        first 100bp 3UTR coverage threshold (default: 10)
  --sample-list sample_list [sample_list ...]
                        sample list (default: None)
  -o output_folder, --outdir output_folder
                        output folder (default: DaPars_results)

filter parameters:
  parameters used in DaPars2 filter analysis, this step will be executed only --treat-index and --control-index are specified

  --treat-index TREAT_INDEX [TREAT_INDEX ...]
                        treat sample index, 1 based (default: None)
  --control-index CONTROL_INDEX [CONTROL_INDEX ...]
                        control sample index, 1 based (default: None)
  --3UTR-Coverage-cutoff COVERAGE_CUTOFF
                        3UTR coverage cutoff (default: 30)
  --Num-least-in-treat NUM_LEAST_IN_TREAT
                        Num sample pass 3UTR coverage filter least in treat (default: 1)
  --Num-least-in-control NUM_LEAST_IN_CONTROL
                        Num sample pass 3UTR coverage filter least in control (default: 1)
  --FDR-cutoff FDR_CUTOFF
                        FDR cutoff (default: 0.05)
  --PDUI-cutoff PDUI_CUTOFF
                        PDUI cutoff (default: 0.2)
  --Fold-change-cutoff FOLD_CHANGE_CUTOFF
                        Fold change cutoff (default: 0.58)
```


# Document

The following is the original document

# DaPars (v2.1)

Dynamics analysis of Alternative PolyAdenylation from multiple RNA-seq data

A [Full Documentations](https://github.com/3UTR/DaPars2/wiki) can be found in Wiki page. And a related 3'aQTL pipeline based on Dapars2 can be found [Here](https://github.com/3UTR/3aQTL-pipe).

## Introduction

The dynamic usage of the 3’untranslated region (3’UTR) resulting from alternative polyadenylation (APA) is emerging as a pervasive mechanism for regulating mRNA diversity, stability and translation. Though RNA-seq provides the whole-transcriptome information and a lot of tools for analyzing gene/isoform expression are available, very few tool focus on the analysis of 3’UTR from standard RNA-seq. DaPars v2 is the next generation of DaPars that directly infers the dynamic alternative polyadenylation (APA) usage by comparing standard RNA-seq from multiple samples. Given the annotated gene model, DaPars v2 can infer the de novo proximal APA sites as well as the long and short 3’UTR expression levels. Finally, the dynamic APA usages of each samples will be identified.

![Flowchart](https://farm8.staticflickr.com/7814/46170216185_6e5eb332fb.jpg) 

## Workflow

![Flowchart](https://farm8.staticflickr.com/65535/51154541918_8a63879ed1_k.jpg)


## Citation

*Please cite the following articles if you use DaPars2 in your research*:
* Feng X, Li L, Wagner EJ, Li W; TC3A: The Cancer 3′ UTR Atlas, Nucleic Acids Research, Volume 46, Issue D1, 4 January 2018, Pages D1027–D1030
* Li L^, Huang K^, Gao YP, Cui Y, Wang G, Nathan D, Li YM, Chen YE, Ji P, Peng F, William K, Wagner EJ, Li W. (2021) An atlas of alternative polyadenylation quantitative trait loci contributing to complex trait and disease heritability. Nature Genetics. doi: 10.1038/s41588-021-00864-5. 



## Contact

If you have any comments, suggestions, questions, bug reports, etc, feel free to contact Lei Li (lei.li.bioinfo@gmail.com). And PLEASE attach your command line and log messages if possible.


